### ISSUE TITLE
Warn about static lambda capturing local variable by reference

### ISSUE THREAD

**Issue by HighCommander4** [llvm/llvm-project#111316]
Capturing a local variable by reference in a static lambda is a recipe for undefined behaviour:

```c++
int foo() {
  int result = 0;
  static auto lambda = [&result] {
    ++result;
  };
  lambda();
  return result;
}

int main() {
  foo();
  foo();  // oops, the lambda is modifying the `result` variable from the first call
}
```

It would be nice if clang warned about this.

(Such a bug was caught recently in the LLVM codebase itself, https://github.com/llvm/llvm-project/issues/109367.)

**Comment by shafik** [comment]
Note asan does catch this: https://godbolt.org/z/ahGY31f6r
